test/integration/filters/continue_headers_only_inject_body_filter.cc

```diff
@@ -28,17 +28,42 @@ class ContinueHeadersOnlyInjectBodyFilter : public Http::PassThroughFilter {
     return Http::FilterHeadersStatus::ContinueAndDontEndStream;
   }
 
-  Http::FilterHeadersStatus encodeHeaders(Http::ResponseHeaderMap& headers, bool) override {
+  Http::FilterHeadersStatus encodeHeaders(Http::ResponseHeaderMap& headers,
+                                          bool end_stream) override {
     headers.setContentLength(body_.length());
-    encoder_callbacks_->dispatcher().post([this]() -> void {
+    encoder_callbacks_->dispatcher().post([this, end_stream]() -> void {
+      posted_ = true;
       Buffer::OwnedImpl buffer(body_);
-      encoder_callbacks_->injectEncodedDataToFilterChain(buffer, true);
+      encoder_callbacks_->injectEncodedDataToFilterChain(buffer, end_stream);
+      if (encoded_) {
+        encoder_callbacks_->continueEncoding();
+      }
     });
-    return Http::FilterHeadersStatus::ContinueAndDontEndStream;
+    if (end_stream) {
+      return Http::FilterHeadersStatus::ContinueAndDontEndStream;
+    } else {
+      return Http::FilterHeadersStatus::Continue;
+    }
+  }
+
+  Http::FilterDataStatus encodeData(Buffer::Instance&, bool) override {
+    encoded_ = true;
+    if (!posted_) {
+      return Http::FilterDataStatus::StopIterationAndBuffer;
+    }
+    return Http::FilterDataStatus::Continue;
   }
 
 private:
   constexpr static absl::string_view body_ = "body";
+  // For HTTP/3 upstream, the headers and fin will arrive separately.
```

Comment:
so when we say that HTTP/3 doesn't have header only its because we'd at least expect to see a headers(end_stream=false) with an empty data(end_stream=true)? Seems not great that we have all these differences in callback usage

Reply:
largely these tests make sure you can go from 
headers + fin
to 
headers + body + fin.
for QUIC, we just insert the data at a different part of the pipeline.  I still think it's important to make sure it works.

Reply:
Makes sense! I think I'm just worried about filter authors having to be aware of all these variations in callback patterns, but this isn't something we need to address in this PR